ff/handlers/nonbonded.py

```diff
@@ -161,6 +162,82 @@ def parameterize_ligand(params, param_idxs):
     return params[param_idxs]
 
 
+def compute_or_load_am1_charges(mol):
+    """Unless already cached in mol's "AM1Cache" property, use OpenEye to compute AM1 partial charges."""
+
+    # check for cache
+    cache_key = "AM1Cache"
+    if not mol.HasProp(cache_key):
+        # The charges returned by OEQuacPac is not deterministic across OS platforms. It is known
+        # to be an issue that the atom ordering modifies the return values as well. A follow up
+        # with OpenEye is in order
+        # https://github.com/openforcefield/openff-toolkit/issues/983
+        am1_charges = list(oe_assign_charges(mol, "AM1"))
+
+        mol.SetProp(cache_key, base64.b64encode(pickle.dumps(am1_charges)))
+
+    else:
+        am1_charges = pickle.loads(base64.b64decode(mol.GetProp(cache_key)))
+
+    return np.array(am1_charges)
```

Comment:
Note that `oe_assign_charges(mol, "AM1")` is conformer dependent. Because the cache is stored onboard an object with mutable conformers, it is possible for the cache to become stale.

For example, if the function is called repeatedly between events that modify `mol`'s conformers in-place:

```python
from rdkit import Chem
from rdkit.Chem import AllChem

from ff.handlers.nonbonded import compute_or_load_am1_charges

# from https://github.com/proteneer/timemachine/blob/a1877dfa2337816c9e8b753f23d558cc05fbb6d9/tests/test_handlers.py#L532
smi = "Clc1c(Cl)c(Cl)c(-c2c(Cl)c(Cl)c(Cl)c(Cl)c2Cl)c(Cl)c1Cl"

def get_mol():
    mol = Chem.MolFromSmiles(smi)
    mol = Chem.AddHs(mol)
    return mol


def precompute_conformers():
    mol = get_mol()
    AllChem.EmbedMultipleConfs(mol)
    return list(mol.GetConformers())[::5]

precomputed_conformers = precompute_conformers()


def set_conformer(mol, i):
    mol.RemoveAllConformers()
    mol.AddConformer(precomputed_conformers[i])

mol_0 = get_mol()
set_conformer(mol_0, 0)
charges_0 = compute_or_load_am1_charges(mol_0)

mol_1 = get_mol()
set_conformer(mol_1, 1)
charges_1 = compute_or_load_am1_charges(mol_1)

print('q(conf_1) - q(conf_0)\n', charges_1 - charges_0)

# modify conformer in place: mol_0 now contains conf_1
set_conformer(mol_0, 1)

# will return outdated charges
expect_charges_1 = compute_or_load_am1_charges(mol_0)
print('q_cached_0(conf_1) - q(conf_0)\n', expect_charges_1 - charges_0) # zeros
```

To protect against this, we might want to modify the caching logic / implementation, so that the computed charges are a deterministic function of the `mol` and `conformer`.

Reply:
I'm not sure I fully understand this, but it sounds like modifying `cache_key` to include a substring that identifies the conformer could be a solution?